### PR TITLE
ASOC: SOF: mediatek: Add ipc support for mt8195

### PR DESCRIPTION
**Pull request by kuanhsuncheng**
This patch adds IPC support for mt8195 using mbox.

On mt8195 resource, there are two mboxes are used to handle
ipc request and reply.
We create a mtk-adsp-ipc client device to request mbox
controllers.

Signed-off-by: Allen-KH Cheng <Allen-KH.Cheng@mediatek.com>

Note: there are three patches delivered upstream
https://lore.kernel.org/linux-arm-kernel/20211028132655.8136-1-allen-kh.cheng@mediatek.com/t/

### CODE REVIEW

File: sound/soc/sof/mediatek/mt8195/mt8195.c
Comment (plbossart) [inline sound/soc/sof/mediatek/mt8195/mt8195.c:356]: where is the platform_device_unregister() done?

File: sound/soc/sof/mediatek/mt8195/mt8195.c
Comment (plbossart) [inline sound/soc/sof/mediatek/mt8195/mt8195.c]: i am not sure I get what you are trying to do here, but the drvdata is typically private stuff handled in the scope of a device. Here you are using it at the parent level. is this intentional and by-design?

**Review by plbossart — COMMENTED** [review]
I don't fully follow what you are trying to do with a platform device here...

File: sound/soc/sof/mediatek/mt8195/mt8195.c
Comment (kuanhsuncheng) [reply sound/soc/sof/mediatek/mt8195/mt8195.c:356]: Sorry, I miss this part, will fix this. thanks.

File: sound/soc/sof/mediatek/mt8195/mt8195.c
Comment (kuanhsuncheng) [reply sound/soc/sof/mediatek/mt8195/mt8195.c]: YES, this is intentional and by-design. private data is used in mtk adsp IPC IRQ callback func.

**Review by kuanhsuncheng — COMMENTED** [review]
I update the  platform_device_unregister part.

For this patch, I refer to imx design.

Note:
Wait upstream done. this patch should not cause compile error
https://lore.kernel.org/linux-arm-kernel/20211028132655.8136-1-allen-kh.cheng@mediatek.com/t/

File: sound/soc/sof/mediatek/mt8195/mt8195.c
Comment (dbaluta) [reply sound/soc/sof/mediatek/mt8195/mt8195.c]: imx design uses a separate driver for IPC communication using the mailbox framework (see drivers/mailbox/imx-mailbox.c and drivers/firmware/imx/imx-dsp.c).

We create a child platform device that will be handled by imx-dsp.c mailbox driver. with ipc_set_data call we pass some information from parent to the child.

I'm not sure if MTK has the mailbox driver though because I havent seen one. 

Later edit: Oh, now I have seen the link for the mailbox driver in @kuanhsuncheng  comment.

@kuanhsuncheng Please Cc me for this kind of patches or at least sound open firmware mailing list. sound-open-firmware@alsa-project.org

Comment (kuanhsuncheng) [reply sound/soc/sof/mediatek/mt8195/mt8195.c]: yes,

There are three patches are waiting for reviewer reply.

mailbox controller driver
https://patchwork.kernel.org/project/linux-mediatek/patch/20211028132756.8251-1-allen-kh.cheng@mediatek.com/

adsp ipc protocol driver (mailbox client interface)
https://patchwork.kernel.org/project/linux-mediatek/patch/20211028132822.8306-1-allen-kh.cheng@mediatek.com/

This PR should wait upstream finished.

Comment (dbaluta) [reply sound/soc/sof/mediatek/mt8195/mt8195.c]: @kuanhsuncheng please mark this PR as draft so that it doesnt get accidentally merged. https://github.blog/changelog/2020-04-08-convert-pull-request-to-draft/

**Comment by cujomalainey** [comment]
Did the patch stall getting sent upstream?

File: sound/soc/sof/mediatek/mt8195/mt8195.c
Comment (kuanhsuncheng) [reply sound/soc/sof/mediatek/mt8195/mt8195.c]: I resend patches again.

(https://lore.kernel.org/linux-arm-kernel/20211117131222.30541-4-allen-kh.cheng@mediatek.com/T/)

Also cc sound-open-firmware@alsa-project.org

Thanks.

**Comment by kuanhsuncheng** [comment]
> Did the patch stall getting sent upstream?

I resend the related patches again.

(https://lore.kernel.org/linux-arm-kernel/20211117131222.30541-4-allen-kh.cheng@mediatek.com/T/)

I think this PR should wait upstream finished.

File: sound/soc/sof/mediatek/mt8195/mt8195.c
Comment (lyakh) [inline sound/soc/sof/mediatek/mt8195/mt8195.c]: you mean `PTR_ERR()`

**Review by lyakh — CHANGES_REQUESTED** [review]
feel free to discard my comment after you've addressed it if I'm still on vacation at the time

**Comment by kuanhsuncheng** [comment]
Hi sirs,

The below is the upstream list for other related patches.

https://patchwork.kernel.org/project/linux-mediatek/list/?series=593601 


I had upstreamed for a while. 

Could someone help me review those patches ?

I would appreciate your help. Many thanks.

**Comment by plbossart** [comment]
> Could someone help me review those patches ?

@kuanhsuncheng it's a bit confusing since this PR is still a 'draft'. It's not clear if reviewers should spent time on this if you don't update the status?

**Comment by kuanhsuncheng** [comment]
> > Could someone help me review those patches ?
> 
> @kuanhsuncheng it's a bit confusing since this PR is still a 'draft'. It's not clear if reviewers should spent time on this if you don't update the status?

Hi @plbossart ,

This PR is still a draft because I am waiting the below patch list upstream done.

https://patchwork.kernel.org/project/linux-mediatek/list/?series=593601

I wonder maybe there is a way can speed up.

**Comment by kuanhsuncheng** [comment]
@lgirdwood , @dbaluta  @plbossart 

Hi sirs,

I wonder is that ok  I take a another PR for below series.

https://patchwork.kernel.org/project/linux-mediatek/list/?series=605114
https://patchwork.kernel.org/project/linux-mediatek/list/?series=605115

and merge this current PR to sof-dev ?


There are other patches for MT8195 needed to merge also.

However, they have a ipc API dependency and cannot directly merge.

I think maintainer may not have time to rely me. (Upstream series=589151)

please kindly give me advice, thanks.

**Comment by plbossart** [comment]
@kuanhsuncheng is this PR still relevant or can it be closed?

**Comment by kuanhsuncheng** [comment]
Hi sirs,

We can close this PR. 

Mark has applied to  https://urldefense.com/v3/__https:/git.kernel.org/pub/scm/linux/kernel/git/broonie/sound.git__;!!CTRNKA9wMg0ARbw!wU0fgArK_q_tKNK7UqFxsnd2931eAre0YciHrlAxzzQaER8C2tTATfYaiNuALVi85laWXg$  for-next

[2/3] ASoC: SOF: mediatek: Add ipc support for mt8195
      commit: aea9350108ed1627f8610c93de44578162b3ee91

Thanks,

**Pull request closed without merging**
